starboard/android/shared/media_codec_bridge_eradicator.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/android/shared/media_codec_bridge_eradicator.h"
+
+#include <pthread.h>
+
+#include "starboard/android/shared/jni_utils.h"
+#include "starboard/common/log.h"
+#include "starboard/common/once.h"
+
+namespace starboard {
+namespace android {
+namespace shared {
+
+SB_ONCE_INITIALIZE_FUNCTION(MediaCodecBridgeEradicator,
+                            MediaCodecBridgeEradicator::GetInstance);
+
+namespace {
+
+struct EradicateParam {
+  explicit EradicateParam(MediaCodecBridgeEradicator* eradicator,
+                          jobject j_media_codec_bridge,
+                          jobject j_reused_get_output_format_result)
+      : eradicator(eradicator),
+        j_media_codec_bridge(j_media_codec_bridge),
+        j_reused_get_output_format_result(j_reused_get_output_format_result) {}
+  MediaCodecBridgeEradicator* eradicator;
+  jobject j_media_codec_bridge;
+  jobject j_reused_get_output_format_result;
+};
+
+}  // namespace
+
+void MediaCodecBridgeEradicator::CheckEradicatingAndMaybeWait() {
+  ScopedLock scoped_lock(mutex_);
+  while (!j_media_codec_bridge_set_.empty()) {
+    condition_variable_.Wait();
+  }
+}
+
+bool MediaCodecBridgeEradicator::Eradicate(
+    jobject j_media_codec_bridge,
+    jobject j_reused_get_output_format_result) {
+  EradicateParam* param = new EradicateParam(this, j_media_codec_bridge,
+                                             j_reused_get_output_format_result);
+
+  // spawn a thread to run the release method on the MediaCodecBridge java
+  // object.
+  pthread_t thread;
+  int result = pthread_create(
+      &thread, nullptr, &MediaCodecBridgeEradicator::EradicateMediaCodecBridge,
+      param);
+  return result == 0;
+}
+
+void* MediaCodecBridgeEradicator::EradicateMediaCodecBridge(void* context) {
+  EradicateParam* param = static_cast<EradicateParam*>(context);
+  SB_DCHECK(param != nullptr);
+
+  MediaCodecBridgeEradicator* eradicator = param->eradicator;
+  jobject j_media_codec_bridge = param->j_media_codec_bridge;
+  jobject j_reused_get_output_format_result =
+      param->j_reused_get_output_format_result;
+
+  {
+    // add the j_media_codec_bridge to the set before the work is started
```

Comment:
This should be added in `Eradicate()` to avoid race condition.

Reply:
My reasons for keeping it here.
1. The core part is the JNI method calls. Putting it here guard the code as minimum as possible, and it will reduce the waiting time of the other thread. 
2. Put in Eradicate() will need to handle the case when the thread is not created successfully. 

If you could explain the race condition better and convince me, then I will move, otherwise, I will keep it here.

Reply:
I do not think this example is valid. 
We use the mutex only to block the creation of the media_codec_bridge java object when the old java object is in the releasing stage.
We do not use it to block the creation of the new media_codec_bridge when there are existing media_codec_bridge objects. 

Even if the code is changed to what you suggested, thread 3 can still run before thread 1, so that thread 3 is creating the new media_codec_bridge object, while thread 2 is destroying the old object.

Reply:
If the code is changed per my comment, thread 3 will call WaitForPendingDestructions() and block until thread 2 destroying the old object.  If we keep the current implementation, thread 3 will still call WaitForPendingDestructions(), but it may not block as thread 2 hasn't insert the bridge to `j_media_codec_bridge_set_`.

Reply:
Assuming these user cases.

Case 1:
Thread 1 run Destroy at time 100
Thread 2 starts at time 101-151
Thread 3 run WaitForPendingDestructions() at time 102
Result - Thread 3 blocks

Case 2:
Thread 3 run WaitForPendingDestructions() at time 100
Thread 1 run Destroy at time 101
Thread 2 starts at time 102-152
Result - Thread 3 runs

Case 3:
Thread 3 run WaitForPendingDestructions() at time 100
Thread 1 run Destroy at time 99
Thread 2 starts at time 101-151
Result - 
If the code is changed per your comment, thread 3 blocks, if not, thread 3 runs.

I do not think the different result for case 3 matters, because of the above 2 cases.

Reply:
Imagine an scenario before this change: on a platform with a single video decoder, the web app stops a playback and then immediately start a new one.  Since the existing implementation destroys the MediaCodec synchronously, this scenario works.

With the change that turns the destruction to be asynchronous, a potential sequence could be:
1. At time 100, the web app stops the first playback, Destroy() called on thread 1, which starts thread 2.
2. At time 110, the web app starts the second playback, WaitForPendingDestructions() is called on thread 3 and returns immediately.  The creation of the MediaCodec fails, which in turn fails the playback.
3. At time 120, thread 2 gets a chance to run and destroys the MediaCodec.

The above scenario fails and may cause regressions in production.

Reply:
Your example is valid, however, my gut feeling is even though there is a regression in your example, it should not be a problem because it already support to create multiple media_coded_bridge instances on production.

I will make the change though.